mmf/modules/losses.py

```diff
@@ -559,3 +558,20 @@ def __init__(self, params=None):
 
     def forward(self, sample_list, model_output):
         return self.loss_fn(model_output["scores"], sample_list.targets)
+
+
+@registry.register_loss("inbatch_cross_entropy")
+class BatchCrossEntropyLoss(nn.Module):
+    def __init__(self, params=None):
+        super().__init__()
+        if params is None:
+            params = {}
+        self.loss_fn = nn.CrossEntropyLoss(**params)
+
+    def forward(self, sample_list, model_output):
+        bs = len(sample_list['id'])
+        scores = model_output['scores']
+        assert scores.shape[0] == scores.shape[1]
+        # our targets are just along the diagonal
+        targets = torch.arange(0, bs, dtype=torch.long).to(device=scores.device)
```

Comment:
This seems to depend only on the batch size. I think this can be done in the model forward pass as well and you can add it to the sample_list.